react/week-2/lesson.md

```diff
@@ -187,276 +188,289 @@ function Counter(props) {
 
 If you look in the console, you'll see that the component is rendered once when the page loads. `props.likeCount` starts at 0, so React inserts "Count: 0" into the DOM.
 
-But when you click the button, the function component is called again (or *re-rendered*). Don't worry about **how** this happens right now. This time `props.likeCount` is **1**. React now **updates** the DOM to make sure it shows the correct number. Every time we click the button, the function component is called and React updates the DOM for us.
+We won't look at how this works at the moment, but behind the scenes there is some code that will listen for clicks on the button and force React to update. That means when you click the button, the function component is called again (or *re-rendered*).
+
+Now `props.likeCount` is **1**. React now **updates** the DOM to make sure it shows the correct number. Every time we click the button, the function component is called and React updates the DOM for us.
 
 We don't need to worry about changing the DOM ourselves! This is what makes React so powerful. Even better, React will figure out exactly the right bits of the DOM that need to be changed, a concept called the ["virtual DOM"](https://reactjs.org/docs/faq-internals.html). This makes it extremely efficient and fast.
 
 ## State
 
-Let's take another look at the the counter example. What if you wanted to create multiple counters on the same page? How would you add another counter?
+*State* is a general concept in software engineering. It is used when a part of your app needs to "remember" something that changes when people interact with it.
```

Comment:
Hey, @40thieves I liked the way you introduce state as a general component concept. I would like to suggest a slightly different way to introduce some of the concepts we are working here. Let me know what do you think.

## State

One of the many strategies that React uses to decide when is the right time to reflect any change on the DOM is to watches a specific set of variables. Those variables are called state. It means that not everything I change inside a component will result in a DOM change? Yeap, that is correct. 

```js
function Counter(props) {
  let counter = 0;

  setTimeout(() => {
     counter = counter += 2;
  }, 1000); 

  return <div>{ counter }</div>;
}
```

What should we expect to see when we run this code and wait for one second? Spoiler! Nothing will happens. Why? The reason why is we are setting a regular variable inside this component, React don't care about regular variables. We could change it to a string, array, object, none of those changes would make this component change from 0 to two in after one second.

So, how do we make React watch to our changes? We have to use a state instead.

**But what is a state?** 

*State* is a general concept in software engineering. It is used when a part of your app needs to "remember" something that changes when people interact with it.

![State diagram](../assets/state-diagram.png)

This is a simple example, but if we had lots of bits of state, then we can make very complex apps.

React provides a function to help use to use states. it is conveniently called useState. We have to import it from React packages.

```js
import React, { useState } from 'react';

console.log(useState)
```

If we look at the console, `useState` is just a function. It lives inside the React code that you installed when you created the app.

To reference the `useState` function in our component, we need to import it from the React code. The curly braces around `useState` are a bit like writing:

```js
let useState = React.useState;
```

In fact we can just write `React.useState` in our component if we want! But to type a bit less code, we import it (using the curly braces) once and then can just use `useState`.

### Using `useState`

Let's refactor the function above to use state rather than `let`.

```js
function Counter(props) {
  const [count, setCount] = useState(0);

  setTimeout(() => {
     setCount(count + 2);
  }, 1000); 

  return <div>{ counter }</div>;
}
```

Reply:
Hi Mozart 🙂 Thanks for the suggestion!

Anthony had (and I adapted) a similar section about the difference between variables and state. The easiest way to see it is in this commit: https://github.com/CodeYourFuture/syllabus/pull/448/commits/ae96a11606d8013a20df38040d5be736f49a8d07

I decided to remove it since Neill pointed out that it may be confusing that the first code example in the state section doesn't actually work. (See the very long feedback thread in Slack). I am also somewhat concerned about the amount of material we have to get through, so decided to cut it.

I could however be persuaded, as we can now point to the cheatsheet for copy/paste code snippets. The syllabus here I think is intended to be mostly read in large chunks, not as a reference guide.

---

If you look at the old class-based state lesson (the easiest way is probably to look at the backed-up scotland syllabus: https://codeyourfuture.github.io/syllabus-scotland/react/week-20/lesson.html#reacting-to-changes) you'll see that I had a similar idea previously - refactoring a broken app to use state.

The feedback generally was that this problematic because:

1. Similar to above, it teaches the "wrong way" first
2. Students generally found my explanation of forced re-rendering confusing

I tried to capture the essence of this in the new Re-Rendering Components section: it demos a similar app, although it does actually work (by re-calling `ReactDOM.render`).